app/src/main/java/org/mozilla/fenix/historymetadata/HistoryMetadataFeature.kt

```diff
@@ -4,78 +4,278 @@
+     *
+     * @param historyHighlights List of history highlights. Can be empty.
+     * @param historyGroups List of history groups. Can be empty.
+     *
+     * @return [RecentlyVisitedItem] list representing the data expected by clients of this feature.
+     */
+    @VisibleForTesting
+    internal fun getCombinedHistory(
+        historyHighlights: List<HistoryHighlightInternal>,
+        historyGroups: List<HistoryGroupInternal>
+    ): List<RecentlyVisitedItem> {
+        // Cleanup highlights now to avoid counting them below and then removing the ones found in groups.
+        val distinctHighlights = historyHighlights
+            .removeHighlightsAlreadyInGroups(historyGroups)
+
+        val totalItemsCount = distinctHighlights.size + historyGroups.size
+
+        return if (totalItemsCount <= MAX_RESULTS_TOTAL) {
+            getSortedHistory(
+                distinctHighlights.sortedByDescending { it.lastAccessedTime },
+                historyGroups.sortedByDescending { it.lastAccessedTime }
+            )
+        } else {
+            var groupsCount = 0
+            var highlightCount = 0
+            while ((highlightCount + groupsCount) < MAX_RESULTS_TOTAL) {
+                if ((highlightCount + groupsCount) < MAX_RESULTS_TOTAL &&
+                    distinctHighlights.getOrNull(highlightCount) != null
+                ) {
+                    highlightCount += 1
                 }
-                .map { (title, data) ->
-                    HistoryMetadataGroup(
-                        title = title,
-                        historyMetadata = data
-                    )
+
+                if ((highlightCount + groupsCount) < MAX_RESULTS_TOTAL &&
+                    historyGroups.getOrNull(groupsCount) != null
+                ) {
+                    groupsCount += 1
                 }
-                .sortedByDescending { it.lastUpdated() }
-                .take(maxResults)
+            }
 
-            homeStore.dispatch(HomeFragmentAction.HistoryMetadataChange(historyMetadata))
+            getSortedHistory(
+                distinctHighlights
+                    .sortedByDescending { it.lastAccessedTime }
+                    .take(highlightCount),
+                historyGroups
+                    .sortedByDescending { it.lastAccessedTime }
+                    .take(groupsCount)
+            )
         }
     }
 
+    /**
+     * Perform an in-memory mapping of a history highlight to metadata records to compute it's last access time.
+     *
+     * - If a `highlight` cannot be mapped to a corresponding `metadata` record, its lastAccessTime will be set to 0.
+     * - If a `highlight` maps to multiple metadata records, its lastAccessTime will be set to the most recently
+     * updated record.
+     *
+     * @param highlights [HistoryHighlight] list for which to get the last accessed time.
+     * @param metadata [HistoryMetadata] list expected to contain the details for all [highlights].
+     *
```

Comment:
nitpick: I think we have more kDoc where we don't separate out the @return with a new line. As a suggestion towards consistency with the codebase, I would defer to opt out of using the new line here.

Reply:
We still have many `@return`s with newlines before and I'd say this has a real advantage - helps with legibility, the same as for `@param`s newlines, one will immediately see these 3 sections in a bigger comment block.